File: seminar-07/dijkstra/python/dijkstra.py
import heapq

INF = 10**18


def dijkstra(adj_list, start):
    n = len(adj_list)
    dist = [INF] * n
    dist[start] = 0

    # в куче храним пары (расстояние, вершина)
    heap = [(0, start)]

    while heap:
        cur_dist, cur = heapq.heappop(heap)

        # если запись устарела — пропускаем
        if cur_dist != dist[cur]:
            continue

        for nxt, w in adj_list[cur]:
            nd = cur_dist + w
            if nd < dist[nxt]:
                dist[nxt] = nd
                heapq.heappush(heap, (nd, nxt))

    return dist

File: seminar-07/dijkstra/cpp/dijkstra.cpp
#include <vector>
#include <queue>
#include <limits>

using namespace std;

using ll = long long;
const ll INF = numeric_limits<ll>::max();

vector<ll> dijkstra(const vector<vector<pair<int, ll>>>& adj, int start) {
    int n = adj.size();
    vector<ll> dist(n, INF);
    dist[start] = 0;

    priority_queue<
        pair<ll,int>,
        vector<pair<ll,int>>,
        greater<pair<ll,int>>
    > pq;

    pq.push({0, start});

    while (!pq.empty()) {
        auto [cur_dist, cur] = pq.top();
        pq.pop();

        if (cur_dist != dist[cur]) continue;

        for (auto [nxt, w] : adj[cur]) {
            ll nd = cur_dist + w;
            if (nd < dist[nxt]) {
                dist[nxt] = nd;
                pq.push({nd, nxt});
            }
        }
    }

    return dist;
}